### PR TITLE
Switch to Typescript

### PR DESCRIPTION
**Pull request by ruediste**
My Anycubic Photon had a hard time with the generated .pws file, so I wanted to add support for .photon. I had a hard time understanding the code and thought some types would greatly help. 

I also dropped the checked in libraries and switched to npm packages. 

Oh, and I also changed the anti-alias count from 0 to 1. UVTools had a division by zero before...

What do you think?

### CODE REVIEW

**Comment by Andrew-Dickinson** [comment]
Oh wow, thanks for the contribution. This is a lot of work. It may take me a bit to review this but I'm absolutely open to making the switch to TS

**Comment by ruediste** [comment]
Great, nice to hear. Looking forward to your feedback.

File: src/renderer/stackup-renderer.ts [hunk 1]
```diff
@@ -1,9 +1,25 @@
+import pcbStackup, { Options } from "pcb-stackup";
 import {SVGToImage} from "./svg_to_png";
+import { GerberFile } from "../ui/utils/pcb_api_utils";
+import gerberToSvg from "gerber-to-svg";
+import whatsThatGerber from "whats-that-gerber";
```
Comment (Andrew-Dickinson) [inline src/renderer/stackup-renderer.ts:5]: Looks like we're getting these transitively through `pcb-stackup`, but if we're using them directly I'd prefer to also list them as direct dependencies in `package.json`

File: src/renderer/pcbAPI.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import renderStackup, { StackupLayer } from "./stackup-renderer"
+import {downloadFiles} from "./download_files"
+import { PrinterModel } from "../ui/export_dialog";
+
+// async function renderStackup(rendered_layers, options){
+//
+// }
+
+/*
+*/
```
Comment (Andrew-Dickinson) [inline src/renderer/pcbAPI.ts]: I think we can clean this up in the migration. No need to keep it around

File: src/ui/file_select_sidebar.tsx [hunk 3]
```diff
@@ -34,62 +37,62 @@ function FileSelectSideBar(props){
...
-               }}/>
+            multiple={true} accept={".g*,.zip,.txt,.xln,.drl"}
+            onInput={() => {
+                if (inputFile.current?.files != null && inputFile.current.files.length > 0) {
+                    setRootFileName("")
+                    setFileList(Array.from(inputFile.current.files))
+                }
+            }} />
+        <input type='file' id='folder' ref={inputFile2} className={"hidden"}
+            onInput={() => {
```
Comment (Andrew-Dickinson) [inline src/ui/file_select_sidebar.tsx]: With these changes we don't seem to be able to upload entire directories anymore. Any particular reason to remove the `directory` and `webkitdirectory` attributes here?

**Review by Andrew-Dickinson — CHANGES_REQUESTED** [review]
Mostly looks good! We still have some `.js` files I see, is that going to cause a problem? (I've never worked in a hybrid environment)

**Comment by ruediste** [comment]
> Mostly looks good! We still have some `.js` files I see, is that going to cause a problem? (I've never worked in a hybrid environment)

No, it should not cause any issues. But you don't have typing for the `.js` files...

File: src/ui/file_select_sidebar.tsx (hunk 3 above)
Comment (ruediste) [reply src/ui/file_select_sidebar.tsx]: Well spotted. It requires some tricking to add the attributes, that's why I did not bother. But they're absolutely required for directory uploads. Fixed.

**Comment by ruediste** [comment]
Any news on this? Waiting for me?

**Comment by Andrew-Dickinson** [comment]
Apologies. I caught COVID. I will try to take a look this weekend

**Review by Andrew-Dickinson — APPROVED** [review]

**Pull request merged**